File: my_av.py
#!/usr/bin/env python3
import math
import string
import re

protocol = "http://"
subdomain = "www"

def length(len1, len2):
    return len1 / len2

def levenshtein_distance(s1, s2):
    m = len(s1)
    n = len(s2)
    d = [[0 for x in range(n + 1)] for y in range(m + 1)]
    max_length = max(len(s1), len(s2))

    for i in range(m + 1):
        d[i][0] = i
    for j in range(n + 1):
        d[0][j] = j

    for j in range(1, n + 1):
        for i in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                d[i][j] = d[i - 1][j - 1]
            else:
                d[i][j] = 1 + min(d[i - 1][j], d[i][j - 1], d[i - 1][j - 1])

    return d[m][n]

def typos(domain):
    domains = ["facebook.com", "chat.whatsapp.com", "en.wikipedia.org",
               "youtube.com", "instagram.com", "paypal.com",
               "itunes.apple.com", "apple.com", "appleid.apple.co",
               "apps.facebook.com", "amazon.com", "google.com",
               "drive.google.com", "groups.google.com", "verify-paypal.com"]
    count = 0
    for i in range(len(domains)):
        if domains[i] in domain:
            if abs(len(domain) - len(domains[i])) > 10:
                return 1
        dist = levenshtein_distance(domain, domains[i])
        if dist >= 1 and dist <= 2:
            return 1
    return 0


def has_malicious_extension(link):
    extension = [".exe", ".dat", ".doc", ".ru", ".ke", ".cc", ".m", ".m68k",
                 ".cl/login", ".bat", ".png", ".pl", ".css", ".arm", ".sh",
                 ".32", ".ru/", "x86", ".co", ".arm7", ".pdf", ".i", ".pm",
                 ".arm6", ".jpg", ".bin", ".dz", ".arm5", ".net", ".pw",
                 "mips", ".mpsl", ".cf", ".br", ".download", ".spc", ".ppc",
                 ".fuku", ".ga", ".org", ".com", "Mozi.m", ".m68k", "/spc",
                 "/gate.php", "/file.php", "/signin", ".csv"]
    for i in range(len(extension)):
        if extension[i] in link:
            pos = len(link) - len(extension[i])
            if link[pos:] == extension[i]:
                return 1
    return 0

def special_character_domain(domain):
    special_characters = [".", "-"]
    count = 0
    for i in range(len(special_characters)):
        count = 0
        for j in range(len(domain)):
            if domain[j] in special_characters[i]:
                count += 1
        if count >= 4 and i == 0:
            return 1
        if count >= 3 and i == 1:
            return 1
    return 0

def special_words_link(link):
    special_words = ["verify", ".login", "security.", "signin", "e=com",
                     "/bin", "admin/"]
    for i in range(len(special_words)):
        if special_words[i] in link:
            return 1
    return 0

def task1():
    malicious = 0
    filename = ""
    link = ""
    domains = []
    domain = ""
    nr_letters = 0
    n = 0
    in_file = open("data/urls/domains_database", "r")
    for line in in_file:
        domains.append(line.strip())
        n += 1
    in_file.close()
    fp = open("data/urls/urls.in", "r")
    if not fp:
        return
    out = open("urls-predictions.out", "w")
    if not out:
        return
    for line in fp:
        malicious = 0
        link = line.strip()
        if has_malicious_extension(link):
            malicious = 1
        if special_words_link(link):
            malicious = 1
        p = link
        if protocol in p:
            p = p[len(protocol):]
        if subdomain in p:
            p = p[len(subdomain):]
            if p[0] != '.':
                malicious = 1
        q = p.find('/')
        if q == -1:
            q = len(link) - 1
        domain = p[:q]
        if special_character_domain(domain):
            malicious = 1
        if typos(domain):
            malicious = 1
        nr_letters = sum(i.isdigit() for i in domain)
        if nr_letters >= 0.35 * len(domain):
            malicious = 1
        for i in range(n):
            if domains[i] in link:
                malicious = 1
        out.write(str(malicious) + '\n')
    fp.close()
    out.close()

def check_bruteforce(data, n):
    sum = 0
    time_string = data[4]
    time_parts = re.split(r'[^0-9]', time_string)
    time_parts = list(filter(None, time_parts))
    days = int(time_parts[0])
    hours = int(time_parts[1])
    mins = int(time_parts[2])
    secs = int(time_parts[3])
    if len(time_parts) > 4:
        milisecs = int(time_parts[4])
        if int(milisecs) >= 1000:
            sum = 1
    sum += 86400.0 * days + 3600.0 * hours + 60 * mins + secs
    if sum > 0.0 and float(data[n - 1]) > 570.0:
        if int(data[5]) > 45 or int(data[6]) > 45:
            return 1
    return 0

def check_cryptominer(data, malign):
    if data[9] == "0" and data[10] == "0" and data[11] == "0":
        return 1
    return 0

def check_safe_ip(data):
    safe_ips = ["255.255.255.255", "ff02::16", "8.8.8.8",
                "239.255.255.250", "36.91.114.86"]
    if data[2] in safe_ips:
        return 1
    return 0

def task2():
    traffic = "data/traffic/traffic.in"
    out = "traffic-predictions.out"
    with open(traffic, "r") as in_file, open(out, "w") as out_file:
        line = in_file.readline()
        for line in in_file:
            malign = 0
            data = line.strip().split(',')
            if check_bruteforce(data, len(data)):
                malign = 1
            if check_cryptominer(data, malign):
                malign = 1
            if check_safe_ip(data):
                malign = 0
            out_file.write(f"{malign}\n")

def main():
    task1()
    task2()

if __name__ == '__main__':
    main()

File: my_av.c
// Copyright Damian Mihai-Robert 312CAb 2022-2023
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <ctype.h>
#include <stdlib.h>

static const char protocol[] = "http://";
static const char subdomain[] = "www";

// Functie care returneaza lungimea unui vector de stringuri
// Am utilizat functia deoarece checkerului nu ii placea altfel
size_t length(size_t len1, size_t len2)
{
	return len1 / len2;
}

// Functie care calculeaza cate diferente sunt intre 2 cuvinte
int levenshtein_distance(const char *s1, const char *s2)
{
	int m = strlen(s1);
	int n = strlen(s2);
	int d[m + 1][n + 1];
	int distance;
	int max_length = fmax(strlen(s1), strlen(s2));

	for (int i = 0; i <= m; i++)
		d[i][0] = i;
	for (int j = 0; j <= n; j++)
		d[0][j] = j;

	for (int j = 1; j <= n; j++) {
		for (int i = 1; i <= m; i++) {
			if (s1[i - 1] == s2[j - 1])
				d[i][j] = d[i - 1][j - 1];
			else
				d[i][j] = 1 + fmin(d[i - 1][j], fmin(d[i][j - 1],
								   d[i - 1][j - 1]));
		}
	}

	return d[m][n];
}

// Functie care verifica daca domeniul este apropiat de unul dintre cele de mai
// jos, caz in care este phishing
int typos(char domain[])
{
	const char domains[][20] = {"facebook.com", "chat.whatsapp.com",
								"en.wikipedia.org", "youtube.com",
								"instagram.com", "paypal.com",
								"itunes.apple.com", "apple.com",
								"appleid.apple.co", "apps.facebook.com",
								"amazon.com", "google.com", "drive.google.com",
								"groups.google.com", "verify-paypal.com"};
	int pos, count;
	size_t len = length(sizeof(domains), sizeof(domains[0]));
	int dist;

	for (int i = 0; i < len; i++) {
		if (strstr(domain, domains[i]))
			if (abs(strlen(domain) - strlen(domains[i])) > 10)
				return 1;
		dist = levenshtein_distance(domain, domains[i]);
		// Daca are 1-2 modificari de nume
		if (dist >= 1 && dist <= 2)
			return 1;
	}
	return 0;
}

// Functie care verifica sa nu am extensii malitioase
int has_malicious_extension(char link[])
{
	const char extension[][55] = {".exe", ".dat", ".doc", ".ke", ".cc",
								  ".m", ".m68k", ".cl/login", ".bat", ".png",
								  ".pl", ".css", ".arm", ".sh", ".32", ".ru/",
								  "x86", ".co", ".arm7", ".pdf", ".i", ".pm",
								  ".arm6", ".jpg", ".bin", ".dz", ".arm5",
								  ".net", ".pw", "mips", ".mpsl", ".cf", ".br",
								  ".download", ".spc", ".ppc", ".fuku", ".ga",
								  ".org", ".com", "Mozi.m", ".m68k", "/spc",
								  "/gate.php", "/file.php", "/signin", ".csv"};
	int pos;
	size_t len = length(sizeof(extension), sizeof(extension[0]));

	for (int i = 0; i < len; i++) {
		if (strstr(link, extension[i])) {
			pos = strlen(link) - strlen(extension[i]);
			if (!strcmp(link + pos, extension[i]))
				return 1;
		}
	}
	return 0;
}

// Functie care verifica sa nu am multe caractere speciale
int special_character_domain(char domain[])
{
	const char special_characters[][2] = {".", "-"};
	int pos, count;
	size_t len;
	len = length(sizeof(special_characters), sizeof(special_characters[0]));

	for (int i = 0; i < len; i++) {
		count = 0;
		for (int j = 0; j < strlen(domain); j++)
			if (strchr(special_characters[i], domain[j]))
				count++;
		// Daca am mai mult de 4 puncte, e cam suspect
		if (count >= 4 && i == 0)
			return 1;
		// Daca am mai mult de 3 liniute, e cam suspect
		if (count >= 3 && i == 1)
			return 1;
	}
	return 0;
}

// Functie care verifica daca linkul are keyworduri specifice malware/phishing
int special_words_link(char link[])
{
	const char special_words[][15] = {"verify", ".login", "security.",
									  "signin", "e=com", "/bin", "admin/"};
	size_t len = length(sizeof(special_words), sizeof(special_words[0]));

	for (int i = 0; i < len; i++) {
		if (strstr(link, special_words[i]))
			return 1;
	}
	return 0;
}

void task1(void)
{
	int malicious = 0;
	char filename[500], link[500], *p, *q, domains[500][500];
	char domain[500];
	int nr_letters, n = 0;
	FILE *in;
	in = fopen("data/urls/domains_database", "r");
	while (fscanf(in, "%s", domain) != EOF) {
		strcpy(domains[n], domain);
		n++;
	}
	fclose(in);
	FILE *fp = fopen("data/urls/urls.in", "r");
	if (!fp)
		return;
	FILE *out = fopen("urls-predictions.out", "w");
	if (!out)
		return;
	while (fscanf(fp, "%s", link) != EOF) {
		malicious = 0;
		if (has_malicious_extension(link))
			malicious = 1;
		if (special_words_link(link))
			malicious = 1;
		p = link;
		if (strstr(p, protocol) == p)
			p = p + strlen(protocol);
		if (strstr(p, subdomain) == p) {
			p = p + strlen(subdomain);
			if (p[0] != '.')
				malicious = 1;
		}
		q = strchr(p, '/');
		if (!q)
			q = link + strlen(link) - 1;
		memcpy(domain, p, q - p);
		domain[q - p] = '\0';
		if (special_character_domain(domain))
			malicious = 1;
		if (typos(domain))
			malicious = 1;
		nr_letters = 0;
		for (int i = 0; i < strlen(domain); i++)
			if (domain[i] >= '0' && domain[i] <= '9')
				nr_letters++;
		if (nr_letters >= 0.35 * strlen(domain))
			malicious = 1;
		for (int i = 0; i < n; i++)
			if (strstr(link, domains[i]))
				malicious = 1;
		fprintf(out, "%d\n", malicious);
	}
	fclose(fp);
	fclose(out);
}

int check_bruteforce(char data[][50], int n)
{
	char *days, *hours, *mins, *secs, *milisecs = NULL;
	float sum;

	days = strndup(data[4], strchr(data[4], ' ') - data[4]);
	hours = strndup(strchr(data[4], ':') - 2, 2);
	mins = strndup(strchr(data[4], ':') + 1, 2);
	secs = strndup(strrchr(data[4], ':') + 1, 2);

	sum = 0;
	if (strrchr(data[4], '.')) {
		milisecs = strdup(strrchr(data[4], '.') + 1);
		if (atoi(milisecs) >= 1000)
			sum = 1;
	}

	sum += 86400.0 * atoi(days) + 3600.0 * atoi(hours) +
			60 * atoi(mins) + atoi(secs);

	free(days); free(hours); free(mins); free(secs);
	if (milisecs)
		free(milisecs);

	// Daca timpul total este mai mare decat 0 si flow_pkts_payload.avg
	// este mare, atunci este bruteforce
	if (sum > 0.0 && atof(data[n - 1]) > 570)
		if ((atoi(data[5]) > 45 || atoi(data[6]) > 45))
			return 1;
	return 0;
}

int check_cryptominer(char data[][50], int malign)
{
	// Daca FIN SYN si ACK sunt 0, este cryptominer
	if (!strcmp(data[9], "0") && !strcmp(data[10], "0") &&
		!strcmp(data[11], "0"))
		return 1;
	return 0;
}

int check_safe_ip(char data[][50])
{
	// Aceste ip-uri sunt ip-uri sigure
	if (!strcmp(data[2], "255.255.255.255") ||
		!strcmp(data[2], "ff02::16") ||
		!strcmp(data[2], "8.8.8.8") ||
		!strcmp(data[2], "239.255.255.250") ||
		!strcmp(data[2], "36.91.114.86"))
		return 1;
	return 0;
}

void task2(void)
{
	FILE *in = fopen("data/traffic/traffic.in", "r");
	FILE *out = fopen("traffic-predictions.out", "w");

	char line[300];
	char data[20][50], *p, *q;
	int n;
	int malign = 0;

	fgets(line, 300, in);
	while (fgets(line, 300, in)) {
		malign = 0;
		n = 0;
		p = strtok(line, ",");
		while (p) {
			strcpy(data[n++], p);
			p = strtok(NULL, ",");
		}

		if (check_bruteforce(data, n))
			malign = 1;

		if (check_cryptominer(data, malign))
			malign = 1;

		if (check_safe_ip(data))
			malign = 0;

		fprintf(out, "%d\n", malign);
	}
	fclose(in);
	fclose(out);
}

int main(void)
{
	task1();
	task2();
	return 0;
}
